packages/core/src/tracing/clsHooked/async_local_storage_context.js

```diff
@@ -0,0 +1,324 @@
+   * or a child context.
+   * @param {string} key
+   * @param {*} value
+   * @returns {Function}
+   */
+  set(key, value) {
+    const context = storage.getStore();
+    if (!context) {
+      throw new Error('No context available. ns.run() or ns.bind() must be called first.');
+    }
+
+    context[key] = value;
+    return unset.bind(null, context, key, value);
+  }
+
+  /**
+   * Retrieves a value by key from the current CLS context (or a parent context), assuming the key/value pair has
+   * been set earlier in this context.
+   * @param {string} key
+   * @param {boolean} [fallbackToSharedContext=false]
+   */
+  get(key, fallbackToSharedContext = false) {
+    const activeContext = storage.getStore() || (fallbackToSharedContext ? this.sharedContext : undefined);
+    if (!activeContext) {
+      return undefined;
+    }
+
+    return activeContext[key];
+  }
+
+  /**
+   * Creates a new CLS context in this namespace.
+   * @returns {InstanaCLSContext}
+   */
+  createContext() {
+    const activeContext = storage.getStore();
+    // Prototype inherit existing context if created a new child context within existing context.
+    const context = Object.create(activeContext || Object.prototype);
+    context._ns_name = this.name;
+    context.id = executionAsyncId();
+    return context;
+  }
+
+  /**
+   * Runs a function in a new CLS context. The context is left after the function terminates. Asynchronous work started
+   * in this function will happen in that new context. The return value from that function (if any) is discarded.
+   * If you aren't 100% certain that the function never returns a value or that client code never relies on that value,
+   * use runAndReturn instead.
+   * @param {Function} fn
+   * @param {InstanaCLSContext} ctx
+   */
+  run(fn, ctx) {
+    const context = ctx || this.createContext();
+
+    storage.run(context, () => fn(context));
+    return context;
+  }
+
+  /**
+   * Runs a function in a new CLS context and returns its return value. The context is left after the function
+   * terminates. Asynchronous work started in this function will happen in that new context.
+   * @param {Function} fn
+   * @param {InstanaCLSContext} ctx
+   */
+  runAndReturn(fn, ctx) {
+    let value;
+    this.run((/** @type {InstanaCLSContext} */ context) => {
+      value = fn(context);
+    }, ctx);
+    return value;
+  }
+
+  /**
+   * Runs a function which returns a promise in a new CLS context and returns said promise. The context is left
+   * as soon as the the promise resolves/is rejected. Asynchronous work started in this promise will happen in that new
+   * context.
+   *
+   * If the given function does not create a then-able, an error will be thrown.
+   *
+   * This function assumes that the returned promise is CLS-friendly or wrapped already.
+   * @param {Function} fn
+   * @param {InstanaCLSContext} ctx
+   */
+  runPromise(fn, ctx) {
```

Comment:
As far as I can see `runPromise` and `runPromiseOrRunAndReturn` **with args** are only called via our SDK from customers? I might be wrong. Wondering if we manually should test https://www.instana.com/docs/ecosystem/node-js/api/#restoring-the-async-context-manually these scenarios. Maybe I have not found where these use cases are covered.

Reply:
- The amqp instrumentation uses `runPromise`: https://github.com/instana/nodejs/blob/edb5d71a51a4c6ec3342e04f2cd30314772084d1/packages/core/src/tracing/instrumentation/messaging/amqp.js#L234
- The bull instrumentation uses `runPromise`:  https://github.com/instana/nodejs/blob/main/packages/core/src/tracing/instrumentation/messaging/bull.js#L182
- Our Lambda wrapper uses `runPromiseOrRunAndReturn`: https://github.com/instana/nodejs/blob/edb5d71a51a4c6ec3342e04f2cd30314772084d1/packages/aws-lambda/src/wrapper.js#L82

Reply:
> with args

:)

Reply:
Discussed: covered in tests